src/lib/operators/join_mpsm/column_materializer_numa.hpp

```diff
@@ -0,0 +1,231 @@
+                                                             std::unique_ptr<PosList>& null_rows_output,
+                                                             ChunkID chunk_id, std::shared_ptr<const Table> input,
+                                                             ColumnID column_id, NodeID numa_node_id) {
+    // This allocator ensures that materialized values are colocated with the actual values.
+    MaterializedValueAllocator<T> alloc{input->get_chunk(chunk_id)->get_allocator()};
+
+    const auto column = input->get_chunk(chunk_id)->get_column(column_id);
+
+    return std::make_shared<JobTask>([this, &output, &null_rows_output, column, chunk_id, alloc, numa_node_id] {
+      resolve_column_type<T>(*column, [&](auto& typed_column) {
+        _materialize_column(typed_column, chunk_id, null_rows_output, (*output)[numa_node_id]);
+      });
+    });
+  }
+
+  /**
+   * Materialization works of all types of columns
+   */
+  template <typename ColumnType>
+  void _materialize_column(const ColumnType& column, ChunkID chunk_id, std::unique_ptr<PosList>& null_rows_output,
+                           MaterializedNUMAPartition<T>& partition) {
+    auto output = std::make_shared<MaterializedChunk<T>>(partition._alloc);
+
+    output->reserve(column.size());
+
+    auto iterable = create_iterable_from_column<T>(column);
+
+    iterable.for_each([&](const auto& column_value) {
+      const auto row_id = RowID{chunk_id, column_value.chunk_offset()};
+      if (column_value.is_null()) {
+        if (_materialize_null) {
+          null_rows_output->emplace_back(row_id);
+        }
+      } else {
+        output->emplace_back(row_id, column_value.value());
+      }
+    });
+
+    if (_sort) {
+      std::sort(output->begin(), output->end(),
+                [](const auto& left, const auto& right) { return left.value < right.value; });
+    }
+
+    partition._chunk_columns[chunk_id] = output;
+  }
+
+  /**
+   * Specialization for dictionary columns
+   */
+  std::shared_ptr<MaterializedChunk<T>> _materialize_column(const DictionaryColumn<T>& column, ChunkID chunk_id,
+                                                            std::unique_ptr<PosList>& null_rows_output,
+                                                            MaterializedValueAllocator<T> alloc) {
+    auto output = MaterializedChunk<T>{alloc};
+    output.reserve(column.size());
+
+    auto value_ids = column.attribute_vector();
+    auto dict = column.dictionary();
+
+    if (_sort) {
+      // Works like Bucket Sort
+      // Collect for every value id, the set of rows that this value appeared in
+      // value_count is used as an inverted index
+
+      auto rows_with_value = std::vector<std::vector<RowID>>(dict->size());
+
+      // Reserve correct size of the vectors by assuming a uniform distribution
+      for (auto& row : rows_with_value) {
+        row.reserve(value_ids->size() / dict->size());
```

Comment:
this will have huge overhead for non-uniform distributions, correct?
EDIT: e.g. when you have one outlier that happens a lot (100x more then the other values) that will pull up the average quite a bit

Reply:
this is actually dead code since we don't support non equi cases. We should think about this problem for the normal sort_merge join though.